### ISSUE TITLE
BUG: stats.bootstrap bca implementation triggers ValueError for `min` statistic

### ISSUE THREAD

**Issue by jcampbell** [scipy/scipy#15883]
### Describe your issue.

`stats.bootstrap` uses a custom implementation of `percentileofscore` that exclusively uses the `strict` behavior from the primary implementation.

As a result, the `min` statistic is always identified as the zeroth percentile, for all datasets. The bca then chooses -inf for z0_hat, which causes a ValueError when the lower percentile is selected as -inf.

I recognize that using bootstrap to estimate confidence intervals of `min` and `max` is not really intended behavior. However, I believe the behavior should be consistent for both of those estimators, and that the `mean` behavior of `percentileofscore` should be used instead.

### Reproducing Code Example

```python
import scipy.stats as stats
import numpy as np
norm = stats.norm.rvs(0, 1, 100)
stats.bootstrap((norm,), statistic=np.min)
```

### Error message

```shell
ValueError                                Traceback (most recent call last)
Input In [1], in <cell line: 4>()
      2 import numpy as np
      3 norm = stats.norm.rvs(0, 1, 100)
----> 4 stats.bootstrap((norm,), statistic=np.min)

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/scipy/stats/_bootstrap.py:454, in bootstrap(data, statistic, vectorized, paired, axis, confidence_level, n_resamples, batch, method, random_state)
    451         return np.percentile(a=a, q=q, axis=-1)
    453 # Calculate confidence interval of statistic
--> 454 ci_l = percentile_fun(theta_hat_b, interval[0]*100)
    455 ci_u = percentile_fun(theta_hat_b, interval[1]*100)
    456 if method == 'basic':  # see [3]

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/scipy/stats/_bootstrap.py:78, in _percentile_along_axis(theta_hat_b, alpha)
     76 for indices, alpha_i in np.ndenumerate(alpha):
     77     theta_hat_b_i = theta_hat_b[indices]
---> 78     percentiles[indices] = np.percentile(theta_hat_b_i, alpha_i)
     79 return percentiles[()]

File <__array_function__ internals>:5, in percentile(*args, **kwargs)

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/numpy/lib/function_base.py:3866, in percentile(a, q, axis, out, overwrite_input, interpolation, keepdims)
   3864 q = asanyarray(q)  # undo any decay that the ufunc performed (see gh-13105)
   3865 if not _quantile_is_valid(q):
-> 3866     raise ValueError("Percentiles must be in the range [0, 100]")
   3867 return _quantile_unchecked(
   3868     a, q, axis, out, overwrite_input, interpolation, keepdims)

ValueError: Percentiles must be in the range [0, 100]
```

### SciPy/NumPy/Python version information

1.7.3 1.21.2 sys.version_info(major=3, minor=9, micro=7, releaselevel='final', serial=0)

**Comment by mdhaber** [comment]
I'm not able to reproduce this `ValueError` issue in `main`.  This is the output I get for the provided example:
```
C:\Users\matth\scipy\scipy\stats\_resampling.py:134: RuntimeWarning: invalid value encountered in double_scalars
  alpha_1 = ndtr(z0_hat + num1/(1 - a_hat*num1))
C:\Users\matth\scipy\scipy\stats\_resampling.py:136: RuntimeWarning: invalid value encountered in double_scalars
  alpha_2 = ndtr(z0_hat + num2/(1 - a_hat*num2))
C:\Users\matth\scipy\scipy\stats\_resampling.py:102: BootstrapDegenerateDistributionWarning: The bootstrap distribution is degenerate; the confidence interval is not defined.
  warnings.warn(BootstrapDegenerateDistributionWarning())
Out[74]: BootstrapResult(confidence_interval=ConfidenceInterval(low=nan, high=nan), standard_error=0.11787258252011637)
```
Is there a seed I should use to get the result you got? It's also possible that the particular error here is already fixed (e.g. by gh-14351).

It sounds like that still does not resolve your issue? You would prefer that `percentileofscore` use a different `kind` option, and I'll have to look into that more closely.

**Comment by jcampbell** [comment]
Indeed -- it appears that #14351 prevents the ValueError by directly checking for the `nan` case. 

That said, in this case I do not think you have a degenerate distribution, and I think the confidence interval should be defined, with the `low` value equal to the sample min. 

The existing `scipy.stats` docs for `percentileofscore` describe the `mean` kind as the one often used in testing, which I think is why that would be more correct in general. Further, as a thought experiment, we should not expect the confidence interval calculation to be impossible when we're checking for the `min` statistic but not for the `max` statistic (which is the case with the current `strict` implementation). So I think I would call this a defect (albeit a minor one) rather than an enhancement.

I could also see the case for `rank` method being the appropriate choice, and that has the benefit of being the default of the current `percentileofscore` implementation. I'm not sure how this change would affect existing tests, but would love to create a simple PR if you agree this is a worthy change.

**Comment by mdhaber** [comment]
> but would love to create a simple PR if you agree this is a worthy change

Sure. I'd be curious to see what changes. I'm sure I originally used `stats.percentileofscore` in the original implementation; I only would have written the custom implementation to improve performance for vectorized input. Could you check how it impacts performance in that case? (For instance, test the example from the documentation "Rather than writing a loop, we can also determine the confidence intervals for all 1000 samples at once.")

It's interesting that you brought this up today. Just this morning I was thinking about gh-15869 and whether something similar would be needed for `bootstrap`. Originally, I thought we would, then this morning I changed my mind because I only remembered how `percentile` was used, forgetting about `percentileofscore`. Now I'm thinking that simply using `percentileofscore` with a different `kind` argument may not be enough; you might also need to include a tolerance factor to avoid missing numerically distinct (but theoretically equal) values in the bootstrap distribution.

**Comment by jcampbell** [comment]
I'll update here with performance notes, but this had no impact on the tests locally, which I found reassuring.

I'm not sure on the tolerance question--my initial thought is that, even with percentileofscore, we are already allowing for that by the simultaneous `<` and `<=` checks, but I have not read https://github.com/scipy/scipy/pull/15892 closely.

**Comment by mdhaber** [comment]
> I'll update here with performance notes, but this had no impact on the tests locally

Yeah when I asked for performance impact I misunderstood the change that would have to be made. The impact is negligible because the comparison is vectorized just like that check that was already there.

> I'm not sure on the tolerance question

The idea in gh-15892 is that exact equality comparison is unreliable. In cases where it might matter, it is better to check e.g. "less than or very close" rather than "less than or equal". I think that idea applies here, but I can take care of it separately.

**Comment by mdhaber** [comment]
I've relabeled this based on the discussion in https://github.com/scipy/scipy/pull/15892.

**Comment by scottshambaugh** [comment]
I just ran into this issue today as well, thanks for putting together a PR!

**Comment by mdhaber** [comment]
@scottshambaugh thanks for posting. Do you have any insight about gh-15892? We were about to merge that, but the original reference does not support the change. Do you know of a reference that would?

**Comment by scottshambaugh** [comment]
@mdhaber unfortunately not, I'm no expert on bootstrapping.

**Comment by mdhaber** [comment]
To summarize, BCa has trouble when the observed statistic is less than or equal to the minimum of the bootstrap distribution. Originally, we saw this with degenerate data (gh-14804), which prompted us to emit the "degenerate data" warning. Now, we see that the same problem can occur in other cases (e.g. degenerate statistic).

I see a few options:
1. Something like gh-15892, if we find out that's OK. I think it is still worth investigating this fix. Even though we ran across this problem in the context of `statistic=np.min`, for which the bootstrap may not be a good idea, the same condition _could_ (in principle) happen for almost any statistic if you get unlucky with the bootstrap samples. The asymmetry of the situation (e.g. `np.min` vs `np.max`) seems unusual, so maybe this is onto something.
2. Study the BCa formulas to see if there is something other than NaN that we should return. The NaN is caused by infinity over infinity, but would it be valid to take a limit here? (I think the confidence interval would collapse to the observed statistic.)
3. Changing the warning message to be less specific about what went wrong; e.g. "The observed value of the statistic is less than or equal to all values in the bootstrap distribution. This is known to occur when, for example, the data is degenerate or the statistic is `np.min`. The BCa confidence interval cannot be calculated."
4. Split the logic that detects the problematic state to distinguish between degenerate data and the more general problem, e.g. "The observed value of the statistic is less than or equal to all values in the bootstrap distribution. The BCa confidence interval cannot be calculated."

Personally, I think it's reasonable to do (3) now while looking into (1). If we find out that (1) is a dead end, we could implement (4) and maybe look into (2).
